### DISCUSSION TITLE
memory used more and more

### DISCUSSION THREAD

**Discussion by toms8421** [scandum/tintin#209] [Bugs]
I use the tt++.exe in windows server,I found that the memory used increase by the time.
at the begining,i create 15 session,the system show me the tt++.exe only used about 15M.after 24h, the memory used about 50M,
72h later,the memory used about 220M.
I try to restart the tt++.exe, when I #zap all active session.the menory not release.until I close the tt++.exe.

**Comment by scandum** [comment]
Is this with the latest version? And are the sessions doing anything specific, or just idling?

**Reply by toms8421** [reply]
yes,it is the latest version 2.02.41.I have some char in a game server. I create a new session for a char to play game with the robot script. when i found that the memory used too big.I discconect all sessions.but the memory used does not decrease immediately.

**Reply by toms8421** [reply]
at the begining,it only use 15M memory,and the chars are doing jobs with the robot script for sever times.I think this memory is enough for the AC,VAR etc. I don't understand why It's increased by times.does any command will alloc memory and the memory should released by another command?

**Comment by scandum** [comment]
Thanks, I'll do some testing. A few thing worth checking:

#var - make sure you're not creating some huge tables
#config buffer_size - make sure this isn't set to 1M or something

Let me know if you find an easy way to reproduce this.

**Reply by toms8421** [reply]
#config buffer_size 1000,
I don't define huge tables for the #var.
but i use the var for flexible tables.
e.g:
#var v_iteminum {};
#ac {  %1(Coin)} {#math v_iteminum[coin] %1 };#nop %1 show me the number of coin.

**Reply by toms8421** [reply]
I have a doubt:when i foud it used 220M memory.I use #zap for all session except #gts, the windows show me that the memory used for tt++.exe also is 220M.

**Reply by scandum** [reply]
Could you try `#info memory` when it gets high again? That'll let me know if it's string memory or some other type.

**Reply by toms8421** [reply]
#info mem
#INFO MEMORY: STACK SIZE: 2428003
#INFO MEMORY: STACK  MAX: 64
#INFO MEMORY: STACK  CAP: 59
#INFO MEMORY: STACK  LEN: 3

#INFO MEMORY: ALLOC SIZE: 4945423
#INFO MEMORY: ALLOC USED: 4700814
#INFO MEMORY: ALLOC  MAX: 262144
#INFO MEMORY: ALLOC  LEN: 155612

#INFO MEMORY: FREED SIZE: 285447
#INFO MEMORY: FREED  MAX: 262144
#INFO MEMORY: FREED  LEN: 34073

#INFO MEMORY: DEBUG SIZE: 1024
#INFO MEMORY: DEBUG  MAX: 30
#INFO MEMORY: DEBUG  LEN: 4

**Reply by toms8421** [reply]
what's the unit for those size number?the task manager show me the tt++.exe used about 221M memory.

**Comment by toms8421** [comment]
when i discconect all session except gts,the #info mem is as follow:
#info mem
#INFO MEMORY: STACK SIZE: 2428003
#INFO MEMORY: STACK  MAX: 64
#INFO MEMORY: STACK  CAP: 59
#INFO MEMORY: STACK  LEN: 3

#INFO MEMORY: ALLOC SIZE: 103788
#INFO MEMORY: ALLOC USED: 9495
#INFO MEMORY: ALLOC  MAX: 262144
#INFO MEMORY: ALLOC  LEN: 155612

#INFO MEMORY: FREED SIZE: 5171477
#INFO MEMORY: FREED  MAX: 262144
#INFO MEMORY: FREED  LEN: 154786

#INFO MEMORY: DEBUG SIZE: 1024
#INFO MEMORY: DEBUG  MAX: 30
#INFO MEMORY: DEBUG  LEN: 4
 
the task manager also show me that tt++.exe occupy about 240M memory.

**Reply by toms8421** [reply]
when i restart the tt++.exe,the mem info is as follow,and the windows task manager show that the tt++.exe has used 20M memory.
> > > #gts
#SESSION 'gts' ACTIVATED.
#info mem
#INFO MEMORY: STACK SIZE: 1128531
#INFO MEMORY: STACK  MAX: 32
#INFO MEMORY: STACK  CAP: 27
#INFO MEMORY: STACK  LEN: 3

#INFO MEMORY: ALLOC SIZE: 4940931
#INFO MEMORY: ALLOC USED: 4705424
#INFO MEMORY: ALLOC  MAX: 262144
#INFO MEMORY: ALLOC  LEN: 143795

#INFO MEMORY: FREED SIZE: 180939
#INFO MEMORY: FREED  MAX: 65536
#INFO MEMORY: FREED  LEN: 22153

#INFO MEMORY: DEBUG SIZE: 1024
#INFO MEMORY: DEBUG  MAX: 30
#INFO MEMORY: DEBUG  LEN: 4
#ses
#THESE SESSIONS HAVE BEEN DEFINED:
kod               47.98.175.3:8000                 (mccp 2)
kodpt             47.98.175.3:8000                 (mccp 2)
kodfc             47.98.175.3:8000                 (mccp 2)
kodmj             47.98.175.3:8000                 (mccp 2)
kodmoon           47.98.175.3:8000                 (mccp 2)
kodmoona          47.98.175.3:8000                 (mccp 2)
kodxs             47.98.175.3:8000                 (mccp 2)
kodtpf            47.98.175.3:8000                 (mccp 2)
kodem             47.98.175.3:8000                 (mccp 2)
kodgm             47.98.175.3:8000                 (mccp 2)
kodlg             47.98.175.3:8000                 (mccp 2)
kodlga            47.98.175.3:8000                 (mccp 2)
kodlgb            47.98.175.3:8000                 (mccp 2)
kodlgc            47.98.175.3:8000                 (mccp 2)
kodlgd            47.98.175.3:8000                 (mccp 2)
kodhyd            47.98.175.3:8000                 (mccp 2)
kodzz             47.98.175.3:8000                 (mccp 2)
koddf             47.98.175.3:8000                 (mccp 2)
kodwdd            47.98.175.3:8000                 (mccp 2)
l
#NO SESSION ACTIVE. USE: #session {name} {host} {port} TO START ONE.

**Comment by scandum** [comment]
The string memory looks fine since the unit size is bytes, only 7MB of memory, so the problem is elsewhere. Does this memory problem happen in version 2.02.32?

**Reply by toms8421** [reply]
I haven't use the version 2.02.32,I use the tt++.exe in this way:
C:\game\game\bin\mintty.exe -c .\bin\mintty.con -i .\bin\tt++.ico -T WinTinjllian -e .\bin\tt++.exe -r main.tin
maybe the program release the memory in code,but doesn't really release it to windows system.such as I discconect all the session except GTS,the windows task manager also show the high memory used.

**Reply by toms8421** [reply]
where is the version 2.02.32? I only find the 2.02.31 in the release version.

**Reply by toms8421** [reply]
#help greeting
      ####################################################################
      #                                                                  #
      #                    T I N T I N + +   2.02.31                     #
      #                                                                  #
      #      Code by Peter Unold, Bill Reis, and Igor van den Hoven      #
      #                                                                  #
      ####################################################################

#info mem
#INFO MEMORY: STACK SIZE: 1318848
#INFO MEMORY: STACK  MAX: 64
#INFO MEMORY: STACK  CAP: 32
#INFO MEMORY: STACK  LEN: 3

#ERROR: index 101581 len = 1/0 max = 282 flags = 6 (8)
#ERROR: index 116042 len = 2/0 max = 372 flags = 6 (11)
#ERROR: index 121909 len = 30/0 max = 282 flags = 6 (se;sw;ne;nw;se;sw;a_search_end)
#INFO MEMORY: ALLOC SIZE: 5004573
#INFO MEMORY: ALLOC USED: 4760148
#INFO MEMORY: ALLOC  MAX: 262144
#INFO MEMORY: ALLOC  LEN: 152752

#INFO MEMORY: FREED SIZE: 210562
#INFO MEMORY: FREED  MAX: 65536
#INFO MEMORY: FREED  LEN: 25408

#INFO MEMORY: DEBUG SIZE: 1024
#INFO MEMORY: DEBUG  MAX: 25
#INFO MEMORY: DEBUG  LEN: 4

I relace the tt++.exe with version 2.02.31, in a few hours,I check the memory used from windows task manager,it's 15M to 29M.

**Reply by scandum** [reply]
My bad, don't think there ever was a .32 release. The error message in `#info memory` should be harmless. Let me know how it goes. :)

**Comment by mbsmith** [comment]
Would it be possible to post what you're running within tintin?
Understand if it's not meant for public viewing, but it might help find
the issue a bit quicker.

**Reply by toms8421** [reply]
does anyone can see the attach file?

**Reply by toms8421** [reply]
[WinTinjllian1.zip](https://github.com/scandum/tintin/files/13904411/WinTinjllian1.zip)

**Comment by mbsmith** [comment]
On Thu, Jan 11, 2024 at 06:19:03AM -0800, toms8421 wrote:
> WinTinjllian1.zip
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.*Message ID: <scandum/tintin/
> ***@***.***>
> 

Yes I see it.  Thanks.

**Comment by toms8421** [comment]
how to build the compiled envioment for tintin within windows ?

**Reply by scandum** [reply]
You can compile tintin under windows with WSL if you run Windows 10 or later.
